zio-http-gen-sbt-plugin/src/main/scala/zio/http/gen/sbt/ZioHttpCodegen.scala

```diff
@@ -0,0 +1,142 @@
+              Def.task {
+                s.log.info("OpenAPI spec unchanged, skipping codegen and using cached files")
+                cached
+              }
+            case _                        =>
+              Def.task {
+                s.log.info("OpenAPI spec changed, or nothing in cache: regenerating code")
+                zioHttpCodegenMakeTask.value
+              }
+          }
+        }
+
+        cachedCodegen(FileInfo.full((ZIOpenApi / sources).value.toSet))
+      }.value,
+    ),
+  ) ++ Seq(
+    Compile / sourceGenerators += (ZIOpenApi / zioHttpCodegenMake).taskValue,
+    Compile / watchSources ++= (ZIOpenApi / sources).value,
+  )
+
+  lazy val zioHttpCodegenMakeTask = Def.task {
+    val openApiFiles: Seq[File]    = (ZIOpenApi / sources).value
+    val openApiRootDirs: Seq[File] = (ZIOpenApi / sourceDirectories).value
+    val baseDir                    = baseDirectory.value
+    val targetDir: File            = (Compile / sourceManaged).value
+    val config: Config             = (ZIOpenApi / zioHttpCodegenConf).value
+
+    openApiFiles.flatMap { openApiFile =>
+      val content        = fileContentAsString(openApiFile)
+      val format         = Format.fromFileName(openApiFile.getName)
+      val openApiRootDir = openApiRootDirs.foldLeft(baseDir) { case (bestSoFar, currentDir) =>
+        val currentPath    = currentDir.getAbsolutePath
+        val isAncestor     = openApiFile.getAbsolutePath.startsWith(currentPath)
+        val isMoreSpecific = currentPath.length >= bestSoFar.getAbsolutePath.length
+        if (isAncestor && isMoreSpecific) currentDir
+        else bestSoFar
+      }
+      val parsedOrError  = format match {
+        case Format.YAML => content.fromYaml[OpenAPI](JsonCodec.jsonDecoder(OpenAPI.schema))
+        case Format.JSON => OpenAPI.fromJson(content)
+      }
+
+      parsedOrError match {
+        case Left(error)    => throw new Exception(s"Failed to parse OpenAPI from $format: $error")
+        case Right(openapi) =>
+          val codegenEndpoints = EndpointGen.fromOpenAPI(openapi, config)
+          val basePackageParts = dirDiffToPackage(openApiRootDir, openApiFile)
+          val currentTargetDir = basePackageParts.foldLeft(targetDir)(_ / _)
+          val currentTargetPat = Path.of(currentTargetDir.toURI())
+
+          CodeGen
+            .renderedFiles(codegenEndpoints, basePackageParts.mkString("."))
+            .map { case (path, content) =>
+              val filePath = currentTargetPat.resolve(path)
+              Files.createDirectories(filePath.getParent)
+              Files.write(filePath, content.getBytes(StandardCharsets.UTF_8), CREATE, TRUNCATE_EXISTING)
+              filePath.toFile
+            }
+            .toSeq
+
+      }
+    }
+  }
+
+  def listFilesRec(dir: File): List[File] =
+    sbt.io.IO.listFiles(dir).foldRight(List.empty[File]) { case (f, acc) =>
+      if (f.isDirectory) listFilesRec(f) ::: acc
+      else f :: acc
+    }
```

Comment:
nit: can be tail recursive
```scala
@tailrec def listFilesRec(dir: File, acc: List[File] = Nil): List[File] =
  sbt.io.IO.listFiles(dir).foldRight(acc) { case (f, tail) =>
    if (f.isDirectory) listFilesRec(f, tail)
    else f :: tail
  }
```

Reply:
this methods should be private?

Reply:
I don't have any strong opinions about it, if you think such stateless methods in plugins are best to limit access to, then OK, I guess.

Reply:
This is not a plugin to recurse through files